### PR TITLE
BUG: groupby().agg fails on categorical column

### LINKED ISSUE #31450: REGR: groupby().agg fails on categorical column in pandas 1.0.0
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
tbl = pd.DataFrame({"col_num": [1, 1, 2, 3]})
tbl["col_cat"] = tbl["col_num"].astype("category")

# The following line works in 0.25.3 but throws exception in 1.0.0
df = tbl.groupby("col_num").agg({"col_cat": "first"})
```
#### Problem description
agg() on categorical column works without any warnings in 0.25.3 but throws exception in pandas 1.0.0:
<details>
NotImplementedError                       Traceback (most recent call last)
<ipython-input-2-663038a86e20> in <module>
----> 1 df = tbl.groupby("col_num").agg({"col_cat": "first"})

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\groupby\generic.py in aggregate(self, func, *args, **kwargs)
    938         func = _maybe_mangle_lambdas(func)
    939 
--> 940         result, how = self._aggregate(func, *args, **kwargs)
    941         if how is None:
    942             return result

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\base.py in _aggregate(self, arg, *args, **kwargs)
    426 
    427                 try:
--> 428                     result = _agg(arg, _agg_1dim)
    429                 except SpecificationError:
    430 

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\base.py in _agg(arg, func)
    393                 result = {}
    394                 for fname, agg_how in arg.items():
--> 395                     result[fname] = func(fname, agg_how)
    396                 return result
    397 

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\base.py in _agg_1dim(name, how, subset)
    377                         "nested dictionary is ambiguous in aggregation"
    378                     )
--> 379                 return colg.aggregate(how)
    380 
    381             def _agg_2dim(name, how):

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\groupby\generic.py in aggregate(self, func, *args, **kwargs)
    245 
    246         if isinstance(func, str):
--> 247             return getattr(self, func)(*args, **kwargs)
    248 
    249         elif isinstance(func, abc.Iterable):

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\groupby\groupby.py in f(self, **kwargs)
   1376                 # try a cython aggregation if we can
   1377                 try:
-> 1378                     return self._cython_agg_general(alias, alt=npfunc, **kwargs)
   1379                 except DataError:
   1380                     pass

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\groupby\groupby.py in _cython_agg_general(self, how, alt, numeric_only, min_count)
    888 
    889             result, agg_names = self.grouper.aggregate(
--> 890                 obj._values, how, min_count=min_count
    891             )
    892 

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\groupby\ops.py in aggregate(self, values, how, axis, min_count)
    579     ) -> Tuple[np.ndarray, Optional[List[str]]]:
    580         return self._cython_operation(
--> 581             "aggregate", values, how, axis, min_count=min_count
    582         )
    583 

~\AppData\Local\Continuum\miniconda3\envs\dev37\lib\site-packages\pandas\core\groupby\ops.py in _cython_operation(self, kind, values, how, axis, min_count, **kwargs)
    453         # are not setup for dim transforming
    454         if is_categorical_dtype(values) or is_sparse(values):
--> 455             raise NotImplementedError(f"{values.dtype} dtype not supported")
    456         elif is_datetime64_any_dtype(values):
    457             if how in ["add", "prod", "cumsum", "cumprod"]:

NotImplementedError: category dtype not supported
</details>
#### Expected Output
no error, just groupped dataframe.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.0.0
numpy            : 1.17.5
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.1.0.post20200119
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.2.7
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.0
IPython          : 7.11.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : 1.2.7
numba            : None
</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [ ] closes #31450 
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
is this limited to "first"?  i think the conclusion from previous thread is that we need to be more systematic about when we call _try_cast

**Comment by charlesdong1991** [comment]
@jbrockmendel yeah, very right, i was about to change something in `_try_cast`, but haven't seen the follow-up discussions, and due to the added change of `isinstance(result[0], dtype.type)` in `_try_cast`, the case of `first` won't be casted back which it should be. 
so I just added as an experiment to see if this could work, and prepare a more generic solution if so.

**Review by WillAyd — CHANGES_REQUESTED** [review]
Hmm not a fan of the approach here - I think should try and avoid special-casing / adding more groupings to base. 

This works for DataFrame selection but not Series right? If so I think should dive deeper into why the former can naturally support but the latter can't to find resolution

**Comment by TomAugspurger** [comment]
Haven't had a chance to look closely yet, but I think my preferred approach is for the calling function to decide what the output dtype should be based on the input dtype. Not sure how best to put that to code either.

**Comment by charlesdong1991** [comment]
@WillAyd @TomAugspurger I agree, and i am not a big fan of this current approach, this is the reason i have [WIP] in the title. I tried to fix when to call this `_try_cast` depending on the situation and type, but not working, and causes a lot of errors locally, so i am still trying to find a better/more robust fix. 

Sorry about this noise in PRs, and i wish i could update this soon!

**Comment by TomAugspurger** [comment]
Thanks for working on this. It's good that we're exploring possible
solutions. We should take a bit of time to try and get this right.

**Comment by charlesdong1991** [comment]
I still have some confusion on what should be the correct behaviour, based on the https://github.com/pandas-dev/pandas/pull/31359#issuecomment-579799671
it seems for `IntegerArray`, `mean` should return a float type result because `IntegerArray._reduce("mean")` returns float. However, in the case below, doing `agg.mean` on a normal array with integer returns `int64` output instead of `float`? is it correct? if so, could this be considered as inconsistency between extension array and normal array and then cause confusion to users? @TomAugspurger @jreback 

```
def test_groupby_extension_agg(self, as_index, data_for_grouping):
        df = pd.DataFrame({"A": [1, 1, 2, 2, 3, 3, 1, 4], "B": data_for_grouping})
        result = df.groupby("B", as_index=as_index).A.mean()
        _, index = pd.factorize(data_for_grouping, sort=True)
    
        index = pd.Index(index, name="B")
        expected = pd.Series([3, 1, 4], index=index, name="A")
        if as_index:
>           self.assert_series_equal(result, expected)
E           AssertionError: Attributes of Series are different
E           
E           Attribute "dtype" are different
E           [left]:  float64
E           [right]: int64
```

**Comment by TomAugspurger** [comment]
IMO the result dtype should only depend on the dtype of the inputs, not the values. IIUC the result can only be int dtype in this case because the mean of the values happens to be equal to an integer? If so, I’d say that’s a bug. The result dtype should always be a float.

**Comment by charlesdong1991** [comment]
@TomAugspurger  thanks for your quick reply. Indeed, in these cases the output has int dtype because it happens to be equal to integer.

**Comment by jreback** [comment]
the int return in mean is a bug it should be float
likely just happened to coerce it before

**Review by charlesdong1991 — COMMENTED** [review]
I think this should pass the CI now, before heading to bed, I would like to describe it a bit, feel free to take a look and your reviews are very welcomed:

1. now the `.agg("mean")` will return `float64`, the same for `var`, `median` etc, it is aligned with numpy behaviour also. And for the rest (defined in the base, they should NOT be casted) and keep the type
2. and `python_agg` will always return the same type as the original object, and it is not changing the current behaviour, since right now those results are coerced. And I think this is correct, because we might want to return different types for `python_agg` and `cython_agg` on cython funcs, and this design is especially good for self-defined function used in aggregate. Therefore, you could see that in test, `.agg("mean")` will have `float64` now while `.agg(lambda x: np.mean(x))` will still return `int64` as they are on master.

I believe this will have a lot of disputes, but all opinions are welcome, and having a consesus on the correct behaviour is important, afterwards I will focus on better coding later. Thanks all, I will do a follow-up code changes tomorrow, and hopefully get this done before 1.0.1 releases. @TomAugspurger @jreback @jbrockmendel @jorisvandenbossche @WillAyd

**Comment by TomAugspurger** [comment]
Thanks for working on this @charlesdong1991. I hope to write up more later, but in general, I like the general idea of the calling function indicating the result dtype is a good idea.

For our immediate needs, I think we can use a more targeted fix to restore the 0.25.x behavior. I'll propose that as a separate PR.

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/pull/31668 for the alternative.

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -807,13 +807,19 @@ def _try_cast(self, result, obj, numeric_only: bool = False):
...
+            # The function can return something of any type, so check
+            #  if the type is compatible with the calling EA.
+            # datetime64tz is handled correctly in agg_series,
+            #  so is excluded here.
             if is_extension_array_dtype(dtype) and dtype.kind != "M":
-                # The function can return something of any type, so check
-                #  if the type is compatible with the calling EA.
-                # datetime64tz is handled correctly in agg_series,
-                #  so is excluded here.
+                from pandas import notna
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: you can import at the top

**Review by jreback — CHANGES_REQUESTED** [review]
thanks for working on this @charlesdong1991 but this is not very clean; i know this is a patch release. but this is making things worse.

**Comment by jreback** [comment]
merged in https://github.com/pandas-dev/pandas/pull/31668, which is a backportable patch. thanks @charlesdong1991

**Comment by charlesdong1991** [comment]
> #31668 for the narrowly scoped alternative to just fix the regression, but not the general issue.

this is much cleaner!! awesome!! @TomAugspurger 👍

**Pull request closed without merging**